libflux/flux-core/src/semantic/infer.rs

```diff
@@ -236,6 +334,19 @@ pub fn instantiate(
                 .collect::<Vec<Constraint>>()
                 .into()
         });
+
+    // Equivalent to `SubstitutionMap` but instantiates bound variables instead of free variables
+    struct InstantiationMap(SubstitutionMap);
```

Comment:
Is this how free vars become bound vars? During instantiation?

Reply:
It is the reverse, we have a `PolyType` with variables bound to it and we instantiate them to free variables.